nuxeo-features/nuxeo-platform-tag-service/nuxeo-platform-tag-core/src/main/java/org/nuxeo/ecm/platform/tag/FacetedTagService.java

```
@@ -0,0 +1,153 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Funsho David
+ */
+
+package org.nuxeo.ecm.platform.tag;
+
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.IdRef;
+import org.nuxeo.ecm.core.versioning.VersioningService;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.nuxeo.ecm.core.api.CoreSession.ALLOW_VERSION_WRITE;
+import static org.nuxeo.ecm.core.query.sql.NXQL.ECM_UUID;
+import static org.nuxeo.ecm.platform.tag.TagConstants.TAG_FACET;
+import static org.nuxeo.ecm.platform.tag.TagConstants.TAG_LIST;
+
+/**
+ * Implementation of the tag service based on facet
+ *
+ * @since 9.3
+ */
+public class FacetedTagService extends AbstractTagService {
+
+    @Override
+    public void doTag(CoreSession session, String docId, String label) {
+        DocumentModel docModel = session.getDocument(new IdRef(docId));
+        if (!docModel.isProxy()) {
+            List<String> tags = getTags(docModel);
+            if (!tags.contains(label)) {
+                tags.add(label);
+            }
+            setTags(docModel, tags);
+            session.saveDocument(docModel);
+        }
+    }
+
+    @Override
+    public void doUntag(CoreSession session, String docId, String label) {
+        if (session.exists(new IdRef(docId))) {
+            DocumentModel docModel = session.getDocument(new IdRef(docId));
```

Why an explicit `exists`? We should just let `getDocument` throw an exception if the docId is invalid.